File: src/common/import_parser.py
"""
Import Parser - Parse @import directives to build tag-to-path mapping

Supports:
1. Single import without 'as':
   @import($__template__.'sessions.tasks')           -> tag: tasks,    path: $__template__.'sessions.tasks'
   @import($__layout__.'base')                        -> tag: base,     path: $__layout__.'base'
   @import('a')                                       -> tag: a,        path: 'a'
   @import("b.d")                                     -> tag: d,        path: "b.d"

2. Single import with 'as':
   @import($__template__.'sessions.projects' as projects) -> tag: projects,  path: $__template__.'sessions.projects'
   @import($__layout__.'base' as baseLayout)              -> tag: baseLayout, path: $__layout__.'base'
   @import($__blade_custom_path__ as alert)               -> tag: alert,      path: $__blade_custom_path__

3. Array import:
   @import([
       'counter' => 'sessions.tasks.count',
       'demo'    => $__template__.'demo.fetch'
   ])
"""

import re
from common.utils import extract_balanced_parentheses


class ImportParser:
    """Parse @import directives and build tag-to-path mapping."""

    def __init__(self):
        self.imports = {}  # tag -> path

    def parse_imports(self, code):
        """
        Parse all @import directives from the code.
        Returns:
            imports: dict mapping tag_name -> raw_path_expression
        """
        self.imports = {}
        
        # Find all @import directives using balanced parentheses
        import_pattern = re.compile(r'@import\s*\(', re.IGNORECASE)
        
        pos = 0
        while pos < len(code):
            match = import_pattern.search(code, pos)
            if not match:
                break
            
            paren_start = match.end() - 1  # Position of '('
            content, end_pos = extract_balanced_parentheses(code, paren_start)
            
            if content is None:
                pos = match.end()
                continue
            
            content = content.strip()
            
            # Remove Blade comments {{-- ... --}} from content
            content = re.sub(r'\{\{--.*?--\}\}', '', content, flags=re.DOTALL).strip()
            
            if content.startswith('['):
                # Array import: @import([ 'tag' => 'path', ... ])
                self._parse_array_import(content)
            else:
                # Single import: with or without 'as'
                self._parse_single_import(content)
            
            pos = end_pos
        
        return self.imports

    def remove_imports(self, code):
        """
        Remove all @import directives from the code.
        Returns cleaned code.
        """
        import_pattern = re.compile(r'@import\s*\(', re.IGNORECASE)
        
        result = code
        # Process in reverse to preserve positions
        matches = []
        pos = 0
        while pos < len(result):
            match = import_pattern.search(result, pos)
            if not match:
                break
            
            paren_start = match.end() - 1
            content, end_pos = extract_balanced_parentheses(result, paren_start)
            
            if content is None:
                pos = match.end()
                continue
            
            matches.append((match.start(), end_pos))
            pos = end_pos
        
        # Remove matches in reverse order
        for start, end in reversed(matches):
            # Also remove trailing Blade comments on same line
            after = result[end:]
            comment_match = re.match(r'\s*\{\{--.*?--\}\}\s*', after, re.DOTALL)
            if comment_match:
                end += comment_match.end()
            # Remove trailing newline
            if end < len(result) and result[end] == '\n':
                end += 1
            result = result[:start] + result[end:]
        
        return result

    def _parse_single_import(self, content):
        """
        Parse a single @import content (non-array).
        
        Formats:
            path_expr                    -> tag from last segment
            path_expr as aliasName       -> tag = aliasName
        """
        content = content.strip()
        
        # Check for 'as' keyword (must be standalone word, not part of path)
        # Pattern: ... as identifier (at the end)
        as_match = re.search(r'\s+as\s+([a-zA-Z_][a-zA-Z0-9_]*)\s*$', content)
        
        if as_match:
            # Has 'as' alias
            tag = as_match.group(1)
            path = content[:as_match.start()].strip()
        else:
            # No 'as' - derive tag from the last segment of the path
            path = content.strip()
            tag = self._extract_tag_from_path(path)
        
        if tag and path:
            self.imports[tag] = path

    def _parse_array_import(self, content):
        """
        Parse array import: [ 'tag1' => 'path1', 'tag2' => path2, ... ]
        """
        # Remove outer brackets
        inner = content.strip()
        if inner.startswith('[') and inner.endswith(']'):
            inner = inner[1:-1].strip()
        
        # Remove Blade comments inside array
        inner = re.sub(r'\{\{--.*?--\}\}', '', inner, flags=re.DOTALL)
        # Remove PHP-style comments // ...
        inner = re.sub(r'//[^\n]*', '', inner)
        
        # Split by commas at level 0 (respecting nested brackets, quotes, parens)
        entries = self._split_at_level_zero(inner, ',')
        
        for entry in entries:
            entry = entry.strip()
            if not entry:
                continue
            
            # Match 'tag' => path or "tag" => path
            arrow_match = re.match(
                r"""['"]([a-zA-Z_][a-zA-Z0-9_]*)['"]  # tag in quotes
                \s*=>\s*                               # =>
                (.+)                                   # path expression
                """,
                entry, re.VERBOSE | re.DOTALL
            )
            
            if arrow_match:
                tag = arrow_match.group(1)
                path = arrow_match.group(2).strip()
                # Remove trailing comma
                path = path.rstrip(',').strip()
                self.imports[tag] = path

    def _extract_tag_from_path(self, path):
        """
        Extract tag name from the last segment of a path expression.
        
        Examples:
            $__template__.'sessions.tasks'  -> tasks
            $__layout__.'base'              -> base
            'a'                             -> a
            "b.d"                           -> d
            $__blade_custom_path__          -> (use variable name heuristic)
        """
        path = path.strip()
        
        # Try to find the last string literal in the path
        # Look for the last quoted string (single or double)
        last_string_match = None
        for m in re.finditer(r"""['"]([^'"]+)['"]""", path):
            last_string_match = m
        
        if last_string_match:
            string_content = last_string_match.group(1)
            # Get the last segment (after the last dot)
            parts = string_content.split('.')
            return parts[-1]
        
        # No string literal found - try to extract from variable name
        # e.g., $__blade_custom_path__ -> blade_custom_path (strip $ and __)
        var_match = re.match(r'\$_*([a-zA-Z][a-zA-Z0-9_]*?)_*$', path)
        if var_match:
            return var_match.group(1)
        
        # Last resort: use the whole path as tag (strip special chars)
        clean = re.sub(r'[^a-zA-Z0-9_]', '', path)
        return clean if clean else None

    def _split_at_level_zero(self, text, delimiter):
        """
        Split text by delimiter, only at nesting level 0.
        Respects brackets [], parentheses (), braces {}, and quotes.
        """
        parts = []
        current = ''
        depth_bracket = 0
        depth_paren = 0
        depth_brace = 0
        in_quote = False
        quote_char = ''
        
        i = 0
        while i < len(text):
            ch = text[i]
            
            if in_quote:
                current += ch
                if ch == quote_char:
                    # Check for escaped quote
                    escape_count = 0
                    j = i - 1
                    while j >= 0 and text[j] == '\\':
                        escape_count += 1
                        j -= 1
                    if escape_count % 2 == 0:
                        in_quote = False
            elif ch in ("'", '"'):
                in_quote = True
                quote_char = ch
                current += ch
            elif ch == '[':
                depth_bracket += 1
                current += ch
            elif ch == ']':
                depth_bracket -= 1
                current += ch
            elif ch == '(':
                depth_paren += 1
                current += ch
            elif ch == ')':
                depth_paren -= 1
                current += ch
            elif ch == '{':
                depth_brace += 1
                current += ch
            elif ch == '}':
                depth_brace -= 1
                current += ch
            elif ch == delimiter and depth_bracket == 0 and depth_paren == 0 and depth_brace == 0:
                parts.append(current)
                current = ''
            else:
                current += ch
            
            i += 1
        
        if current.strip():
            parts.append(current)
        
        return parts

File: src/Template/ImportParser.php
<?php

declare(strict_types=1);

namespace Saola\Compiler\Template;

use Saola\Compiler\Support\Balanced;
use Saola\Compiler\Support\Re;

/**
 * Đọc directive `@import` để dựng bảng THẺ → ĐƯỜNG DẪN.
 *
 * Ba dạng:
 *
 *   @import($__template__.'sessions.tasks')                 → tasks
 *   @import($__layout__.'base' as baseLayout)               → baseLayout
 *   @import(['counter' => 'a.b', 'demo' => $__template__.'c'])
 *
 * Port từ compiler/src/common/import_parser.py.
 */
final class ImportParser
{
    /** @var array<string, string> thẻ → biểu thức đường dẫn nguyên văn */
    private array $imports = [];

    /**
     * @return array<string, string>
     */
    public function parseImports(string $code): array
    {
        $this->imports = [];

        $scan = self::maskVerbatim($code);
        $pos = 0;
        $length = strlen($code);

        while ($pos < $length) {
            if (! Re::match('/@import\s*\(/i', substr($scan, $pos), $m, PREG_OFFSET_CAPTURE)) {
                break;
            }

            $matchStart = $pos + $m[0][1];
            $matchEnd = $matchStart + strlen($m[0][0]);
            $parenStart = $matchEnd - 1;

            // Quét trên bản ĐÃ CHE nhưng cắt chuỗi trên code THẬT — mask giữ
            // nguyên độ dài nên hai chỉ số luôn khớp nhau.
            [$content, $endPos] = Balanced::extractParensAt($code, $parenStart);

            if ($content === null) {
                $pos = $matchEnd;
                continue;
            }

            $content = trim(Re::replace('/\{\{--.*?--\}\}/s', '', trim($content)));

            if (str_starts_with($content, '[') || str_starts_with($content, '{')) {
                $this->parseArrayImport($content);
            } else {
                $this->parseSingleImport($content);
            }

            $pos = $endPos;
        }

        return $this->imports;
    }

    /** Gỡ mọi directive `@import` khỏi code. */
    public function removeImports(string $code): string
    {
        $scan = self::maskVerbatim($code);

        /** @var list<array{int, int}> $matches */
        $matches = [];
        $pos = 0;
        $length = strlen($code);

        while ($pos < $length) {
            if (! Re::match('/@import\s*\(/i', substr($scan, $pos), $m, PREG_OFFSET_CAPTURE)) {
                break;
            }

            $matchStart = $pos + $m[0][1];
            $matchEnd = $matchStart + strlen($m[0][0]);

            [$content, $endPos] = Balanced::extractParensAt($code, $matchEnd - 1);

            if ($content === null) {
                $pos = $matchEnd;
                continue;
            }

            $matches[] = [$matchStart, $endPos];
            $pos = $endPos;
        }

        $result = $code;

        // Gỡ từ CUỐI ngược lên để chỉ số phía trước không bị xê dịch
        foreach (array_reverse($matches) as [$start, $end]) {
            // Kèm luôn comment Blade nằm cùng dòng phía sau
            if (Re::match('/^\s*\{\{--.*?--\}\}\s*/s', substr($result, $end), $c)) {
                $end += strlen($c[0]);
            }

            if ($end < strlen($result) && $result[$end] === "\n") {
                $end++;
            }

            $result = substr($result, 0, $start) . substr($result, $end);
        }

        return $result;
    }

    /**
     * `@import` nằm trong `@verbatim` là VĂN BẢN MẪU, không phải directive.
     *
     * Che vùng đó bằng khoảng trắng CÙNG ĐỘ DÀI để offset vẫn khớp code gốc —
     * nhờ vậy chỉ cần quét trên bản che, còn cắt chuỗi thì dùng code thật.
     */
    private static function maskVerbatim(string $code): string
    {
        return Re::replaceCallback(
            '/@verbatim\b.*?@endverbatim\b/si',
            static fn (array $m): string => str_repeat(' ', strlen($m[0])),
            $code,
        );
    }

    /** `path_expr` hoặc `path_expr as aliasName`. */
    private function parseSingleImport(string $content): void
    {
        $content = trim($content);

        if (Re::match('/\s+as\s+\$?([a-zA-Z_][a-zA-Z0-9_]*)\s*$/', $content, $m, PREG_OFFSET_CAPTURE)) {
            $tag = $m[1][0];
            $path = trim(substr($content, 0, $m[0][1]));
        } else {
            $path = $content;
            $tag = self::extractTagFromPath($path);
        }

        // Python: `if tag and path` — '' và None đều falsy
        if ($tag !== null && $tag !== '' && $path !== '') {
            $this->imports[$tag] = $path;
        }
    }

    /** `[ 'tag1' => 'path1', 'tag2' => path2, ... ]` */
    private function parseArrayImport(string $content): void
    {
        $inner = trim($content);

        if (
            (str_starts_with($inner, '[') && str_ends_with($inner, ']'))
            || (str_starts_with($inner, '{') && str_ends_with($inner, '}'))
        ) {
            $inner = trim(substr($inner, 1, -1));
        }

        $inner = Re::replace('/\{\{--.*?--\}\}/s', '', $inner);
        $inner = Re::replace('/\/\/[^\n]*/', '', $inner);

        foreach (self::splitAtLevelZero($inner, ',') as $entry) {
            $entry = trim($entry);

            if ($entry === '') {
                continue;
            }

            if (Re::match('/^(?:[\'"]?([a-zA-Z_][a-zA-Z0-9_]*)[\'"]?)\s*(?:=>|:)\s*(.+)/s', $entry, $m)) {
                $this->imports[$m[1]] = trim(rtrim(trim($m[2]), ','));
            }
        }
    }

    /**
     * Tên thẻ suy từ đoạn cuối của biểu thức đường dẫn.
     *
     *   $__template__.'sessions.tasks'  → tasks
     *   'a'                             → a
     *   "b.d"                           → d
     *   $__blade_custom_path__          → blade_custom_path
     *
     * PHẢI trùng luật với {@see ImportAliases::deriveName} phía preprocessor —
     * tên thẻ và alias mà trỏ hai nơi khác nhau là hỏng.
     */
    public static function extractTagFromPath(string $path): ?string
    {
        $path = trim($path);

        $literals = Re::matchAll('/[\'"]([^\'"]+)[\'"]/', $path);

        if ($literals !== []) {
            $last = $literals[count($literals) - 1][1];
            $parts = explode('.', $last);

            return $parts[count($parts) - 1];
        }

        if (Re::match('/^\$_*([a-zA-Z][a-zA-Z0-9_]*?)_*$/', $path, $m)) {
            return $m[1];
        }

        // Nhánh cuối trả null (không phải '') khi không còn ký tự hợp lệ —
        // khớp `return clean if clean else None` của bản Python.
        $clean = Re::replace('/[^a-zA-Z0-9_]/', '', $path);

        return $clean === '' ? null : $clean;
    }

    /**
     * Tách ở mức lồng 0, tôn trọng `[]`, `()`, `{}` và nháy (có đếm escape).
     *
     * Biến thể riêng của import_parser.py — khác ba biến thể trong
     * {@see Balanced} ở chỗ đếm số dấu `\` liên tiếp để biết nháy có bị escape
     * thật hay không. Giữ private vì chỉ chỗ này dùng.
     *
     * @return list<string>
     */
    private static function splitAtLevelZero(string $text, string $delimiter): array
    {
        $parts = [];
        $current = '';
        $bracket = 0;
        $paren = 0;
        $brace = 0;
        $inQuote = false;
        $quoteChar = '';

        for ($i = 0, $n = strlen($text); $i < $n; $i++) {
            $ch = $text[$i];

            if ($inQuote) {
                $current .= $ch;

                if ($ch === $quoteChar) {
                    $escapes = 0;
                    for ($j = $i - 1; $j >= 0 && $text[$j] === '\\'; $j--) {
                        $escapes++;
                    }

                    if ($escapes % 2 === 0) {
                        $inQuote = false;
                    }
                }

                continue;
            }

            if ($ch === "'" || $ch === '"') {
                $inQuote = true;
                $quoteChar = $ch;
                $current .= $ch;
                continue;
            }

            match ($ch) {
                '[' => $bracket++,
                ']' => $bracket--,
                '(' => $paren++,
                ')' => $paren--,
                '{' => $brace++,
                '}' => $brace--,
                default => null,
            };

            if ($ch === $delimiter && $bracket === 0 && $paren === 0 && $brace === 0) {
                $parts[] = $current;
                $current = '';
                continue;
            }

            $current .= $ch;
        }

        if (trim($current) !== '') {
            $parts[] = $current;
        }

        return $parts;
    }
}
